compiler/luci/service/src/Nodes/CircleConcatenation.test.cpp

```diff
@@ -47,3 +49,58 @@ TEST(CloneNodeTest, clone_Concatenation_NEG)
   auto cloned = luci::clone_node(node_concat, gc.get());
   ASSERT_EQ(nullptr, cloned);
 }
+
+TEST(ShapeRuleTest, concat_dynamic_shape)
+{
+  luci::CircleInput input_1;
+  luci::CircleInput input_2;
+  luci::CircleConcatenation concat(2);
+
+  input_1.shape({1, 4, 3, 1});
+  input_1.shape_status(luci::ShapeStatus::VALID);
+
+  input_2.shape({1, 4, 3, 1});
+  input_2.shape_status(luci::ShapeStatus::VALID);
+  input_2.dim(2).unset();
+
+  concat.values(0, &input_1);
+  concat.values(1, &input_2);
+  concat.axis(2);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_TRUE(shape_inf_rule.infer(&concat, shape));
+  ASSERT_EQ(4, shape.rank());
+  ASSERT_TRUE(shape.dim(0).known());
+  ASSERT_TRUE(shape.dim(1).known());
+  ASSERT_FALSE(shape.dim(2).known());
+  ASSERT_TRUE(shape.dim(3).known());
+  ASSERT_EQ(1, shape.dim(0).value());
+  ASSERT_EQ(4, shape.dim(1).value());
+  ASSERT_EQ(0, shape.dim(2).value());
+  ASSERT_EQ(1, shape.dim(3).value());
+}
+
+TEST(ShapeRuleTest, concat_rank_mismatch_NEG)
```

Comment:
How about add NEG test where same shape but unknown dim exists at a dimension that is not `axis` of concat?

Reply:
I've added a new test `concat_dynamic_shape_non_axis`. PTAL

Reply:
Ah, the test is not a negative test, because it is a natural flow. Instead, I've added another negative test `concat_wrong_shape_NEG`